crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs

```diff
@@ -0,0 +1,775 @@
+            let param = params.items().iter().next()?.ok()?;
+            let formal_param = param.as_any_js_formal_parameter()?;
+            formal_param.as_js_formal_parameter()?.binding().ok()
+        }
+        AnyJsFunction::JsArrowFunctionExpression(arrow) => {
+            let params = arrow.parameters().ok()?;
+            match params {
+                biome_js_syntax::AnyJsArrowFunctionParameters::JsParameters(js_params) => {
+                    let param = js_params.items().iter().next()?.ok()?;
+                    let formal_param = param.as_any_js_formal_parameter()?;
+                    formal_param.as_js_formal_parameter()?.binding().ok()
+                }
+                biome_js_syntax::AnyJsArrowFunctionParameters::AnyJsBinding(binding) => {
+                    Some(AnyJsBindingPattern::AnyJsBinding(binding))
+                }
+            }
+        }
+        _ => None,
+    }
+}
+
+fn check_root_scope_destructuring(
+    setup_fn: &SetupFunction,
+    props_binding: &JsIdentifierBinding,
+    model: &SemanticModel,
+) -> Vec<Violation> {
+    let mut violations = Vec::new();
+
+    for reference in model.as_binding(props_binding).all_reads() {
+        if let Some(identifier) =
+            biome_js_syntax::JsReferenceIdentifier::cast(reference.syntax().clone())
+            && is_reference_in_root_scope_of_function(&identifier, setup_fn)
+            && let Some(destructuring_info) = is_reference_in_destructuring(&identifier)
+            && !is_safe_reactive_destructuring(&destructuring_info, model)
+        {
+            violations.push(Violation::RootScopeDestructuring {
+                destructuring_range: destructuring_info.destructuring_range,
+                props_param_range: props_binding.range(),
+            });
+        }
+    }
+
+    violations
+}
+
+fn is_reference_in_root_scope_of_function(
+    reference: &biome_js_syntax::JsReferenceIdentifier,
+    function: &SetupFunction,
+) -> bool {
+    let function_syntax = match function {
+        SetupFunction::Function(func) => func.syntax(),
+        SetupFunction::Method(method) => method.syntax(),
+    };
+
+    let nearest_function = reference.syntax().ancestors().find(is_function_like_node);
+
+    nearest_function.is_some_and(|func_node| func_node == *function_syntax)
+}
+
+fn is_function_like_node(node: &biome_rowan::SyntaxNode<biome_js_syntax::JsLanguage>) -> bool {
+    use biome_js_syntax::JsSyntaxKind;
+    matches!(
+        node.kind(),
+        JsSyntaxKind::JS_FUNCTION_DECLARATION
+            | JsSyntaxKind::JS_FUNCTION_EXPRESSION
+            | JsSyntaxKind::JS_ARROW_FUNCTION_EXPRESSION
+            | JsSyntaxKind::JS_METHOD_CLASS_MEMBER
+            | JsSyntaxKind::JS_METHOD_OBJECT_MEMBER
+            | JsSyntaxKind::JS_GETTER_CLASS_MEMBER
+            | JsSyntaxKind::JS_GETTER_OBJECT_MEMBER
+            | JsSyntaxKind::JS_SETTER_CLASS_MEMBER
+            | JsSyntaxKind::JS_SETTER_OBJECT_MEMBER
+    )
+}
```

Comment:
I believe we already have a node for this, but I might be wrong. Check for `AnyFunctionLike` or something similar